api/app/Models/User.php

```diff
@@ -159,28 +186,244 @@ protected function generateSubject()
-        is_null($this->attributes['would_accept_temporary'])
-        )   {
-            return false;
+        // Pool acts as an OR filter. The query should return candidates in ANY of the pools.
+        $query->whereExists(function ($query) use ($pools) {
+            $query->select('id')
+                  ->from('pool_candidates')
+                  ->whereColumn('pool_candidates.user_id', 'users.id')
+                  ->whereIn('pool_candidates.pool_id', $pools);
+        });
+        return $query;
+    }
+    public function filterByLanguageAbility(Builder $query, ?string $languageAbility): Builder
+    {
+        // If filtering for a specific language the query should return candidates of that language OR bilingual.
+        $query->where(function($query) use ($languageAbility) {
+            $query->where('language_ability', $languageAbility);
+            if ($languageAbility == ApiEnums::LANGUAGE_ABILITY_ENGLISH || $languageAbility == ApiEnums::LANGUAGE_ABILITY_FRENCH) {
+                $query->orWhere('language_ability', ApiEnums::LANGUAGE_ABILITY_BILINGUAL);
             }
-        else{
-            return true;
+        });
+        return $query;
+    }
+    public function filterByOperationalRequirements(Builder $query, ?array $operationalRequirements): Builder
+    {
+        // if no filters provided then return query unchanged
+        if (empty($operationalRequirements)) {
+            return $query;
+        }
+
+        // OperationalRequirements act as an AND filter. The query should only return candidates willing to accept ALL of the requirements.
+            $query->whereJsonContains('accepted_operational_requirements', $operationalRequirements);
+        return $query;
+    }
+    public function filterByLocationPreferences(Builder $query, array $locations): Builder
+    {
+        // LocationPreferences acts as an OR filter. The query should return candidates willing to work in ANY of the locations.
+        $query->where(function($query) use ($locations) {
+            foreach($locations as $index => $location) {
+                if ($index === 0) {
+                    // First iteration must use where instead of orWhere
+                    $query->whereJsonContains('location_preferences', $location);
+                } else {
+                    $query->orWhereJsonContains('location_preferences', $location);
+                }
+            }
+        });
+        return $query;
+    }
+    public function filterByJobLookingStatus(Builder $query, ?string $jobLookingStatus): Builder
+    {
+        $query->where('job_looking_status', $jobLookingStatus);
+        return $query;
+    }
+    public function filterBySkills(Builder $query, array $skills): Builder
+    {
+        if (empty($skills)) {
+            return $query;
+        }
+
+        // skills act as an AND filter. The query should only return candidates with ALL of the skills.
+        $query->whereExists(function ($query) use ($skills) {
```

Comment:
I'm impressed! This looks like it took some time to figure out :P It would take me some to just read through it and understand how it works.
Out of curiousity - am I correct that this only works because our ids are UUIDs? If they were non-unique integers, matching skills to experiences (this line `$join->on('experience_skills.experience_id', '=', 'experiences.experience_id');`) would be more complicated?